lib/facter/currentuser.rb

```diff
@@ -0,0 +1,60 @@
+require 'facter'
+
+## Define our Facts as nil first
+username = nil
+name = nil
+domain = nil
+sid = nil
+
+## Fact for full username of the currently/recently logged in user... usually helpful for Endpoints
+Facter.add("windows_currentuser_username") do
+  confine :kernel => 'windows'
+  setcode do
+    require 'win32ole'
+    wmi = WIN32OLE.connect('winmgmts://./root/CIMV2')
+    ## First find out
+    query1 = "select Username from win32_computersystem"
+    wmi.ExecQuery(query1).each do |data|
+      if data.Username.nil?
+        username = nil
+      else
+        domain, name = data.Username.split("\\")
+        username = data.Username
+      end
+
+    end
+    if username.nil?
+      sid = nil
+    else
+      query2 = "select sid from win32_useraccount where name='"+ name +"' and domain='"+ domain +"'"
+      wmi.ExecQuery(query2).each do |data|
+        sid = data.SID
+      end
+    end
+    username

```

Comment:
Facter has a core `id` fact that should return the `"domain\\user"` style name, or the short name for built-in accounts. You could just update the registry provider to do the conversion to sid by calling https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/sid.rb#L16

One thing we've been bitten before is using WMI to query account names and not restricting the query to local accounts.